Gathers & memory issues

I'm opening an anecdotal reporting thread that may hopefully inform usability improvements and bug fixes if/when found. I hope to get more specific as I uncover more & will continue to add info.

In data-heavy dynamic task-of-task workflows, where tasks may return pandas DataFrames on the order of a few GB, users will sometimes create task graphs where one worker client task tries to gather several tasks from many other worker clients (having yielded from the thread pool to gather), but where the total data footprint of the resulting gather will be well into the tens of GBs. Including overhead, this can overwhelm the receiving worker, and start cascading failures because of memory exhaustion.

Ignoring for the moment any design flaws in the program formulation, there are some strange observable behaviors worth reporting:

1 - the nicest and most clear thing that happens is a KilledWorker error, identifying the task that OOM'd the worker N times (for N = allowed-failures+1)

2 - using a LocalCluster w/o a memory limit, I think the nanny dies too because it's co-located within the worker process when the OS (Linux in my case) kills the proc? Anecdotally, what happens SOMETIMES is that tasks are seen to be cancelled, with none found in an error state. Almost always some sort of FATAL-level logging regarding key not found errors are reported by distributed.

3 - in the worst case, a gather thinks it's actually *succeeded*, but at least one of the gathered results is garbled: it comes back either a string instead of a dataframe, or a malformed dataframe.  To see what was happening, I set up one of these degenerate programs to pickle the resulting dataframe from gather to disk when I identified it as garbled (I did this by checking a property I know to be True, specifically if the index is_monotonic_increasing, which showed up as False). Trying to read back the dumped pickle, however, leads to "UnpicklingError: pickle exhausted before end of frame", further indicating data corruption. This is quite hard to reproduce even in my "real" example.

I have yet to produce an example I can share, but I wanted to start a conversation around the observations.

Here's a more detailed trace of one weirdly behaved run. I wrote a scheduler plugin to log when workers leave (which in my case, are because they are memory-killed), and also when tasks transition to an 'erred' state. 

The following comes from a local cluster with six cores and allowed failures = 2. 

My versions btw are distributed 1.21.8 / dask 0.17.4

```
... first, several workers will be killed, having run out of memory. (the number prior to the log level, btw, is the process pid)

2018-06-04 17:08:00,806 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 1>', 'op': 'lost-data'}
2018-06-04 17:08:00,810 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 2>', 'op': 'lost-data'}
2018-06-04 17:08:00,812 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 2>', 'op': 'cancelled-key'}
2018-06-04 17:08:00,813 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 3>', 'op': 'lost-data'}
2018-06-04 17:08:00,815 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 4>', 'op': 'lost-data'}
2018-06-04 17:08:00,817 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 4>', 'op': 'cancelled-key'}
2018-06-04 17:08:00,819 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 5>', 'op': 'lost-data'}
2018-06-04 17:08:00,826 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 6>', 'op': 'lost-data'}
2018-06-04 17:08:00,828 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 6>', 'op': 'cancelled-key'}
2018-06-04 17:08:00,831 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'key': '<task id 7>', 'op': 'lost-data'}
2018-06-04 17:08:00.848 17567 [WARNING] my_plugin:703 | [SCHEDULER] Worker tcp://x.x.x.x:36932 was removed from cluster

... this repeats several times as workers are killed ... 

... finally, i see a fatal log message saying showing a write task-erred op failed ...  but nothing is transitioning to 'erred' state in my plugin!! so my user code never sees this error...
	
2018-06-04 17:11:11,247 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'traceback': None, 'key': '<task id 8>', 'exception': b'\x80\x03cdistributed.scheduler\nKilledWorker...', 'op': 'task-erred'}
2018-06-04 17:11:11,248 17567 [FATAL] distributed.scheduler scheduler.py:1864 | Tried writing to closed comm: {'traceback': None, 'key': '<task id 9>', 'exception': b'\x80\x03cdistributed.scheduler\nKilledWorker...', 'op': 'task-erred'}
2018-06-04 17:11:11.254 17567 [WARNING] my_plugin:703 | [SCHEDULER] Worker tcp://x.x.x.x:36760 was removed from cluster

... things keep chugging along. finally (about 20s later in this case), one of my task results is observed as corrupted - the following error is raised in my code. the dask/distributed error above went unobserved. btw, i haven't inspected the corrupted value yet ... but I would bet on it being a task id instead of a task result. this error also seems to get swallowed.
	
2018-06-04 17:11:33.880 17567 [WARNING] my_plugin:699 | [SCHEDULER] Error #1 observed in task <task id 10> on tcp://x.x.x.x:55579: Logic error: result of type <class 'str'> is not DataFrame

... now, another 50 seconds or so later in some worker, i observe that I haven't gotten all my gathered tasks as expected. this is a gather with skipping errors.  I then check for any tasks where task.status == 'error', (which i should have observed much earlier). and yet I still find none in an error state (all ungathered tasks in cancelled state??!!)

2018-06-04 17:12:21.191 18846 [ERROR] my_business_logic:147 | <<< [FAILURE] gathered 10 results < 14 expected in <task id 11>
2018-06-04 17:12:21.192 18846 [ERROR] my_business_logic:195 | ---- NO TASK ERROR(S)s FOUND
```

--- a/distributed/client.py
+++ b/distributed/client.py
@@ -1461,7 +1461,7 @@ class Client(Node):
                     else:
                         raise ValueError("Bad value, `errors=%s`" % errors)
 
-            keys = [k for k in keys if k not in bad_keys]
+            keys = [k for k in keys if k not in bad_keys and k not in data]
 
             if local_worker:  # look inside local worker
                 data.update({k: local_worker.data[k]

